Set Up Funtions

mem48 opened the pull request:
A new file otp-setup.R containing 4 new functions

otp_build_graph - builds an OPT graph
otp_setup - starts a local OPT server
otp_stop - stops a local OPT server 
otp_checks - runs some basic checks, do we have the correct version of Java installed etc. Called internally by other above functions

The idea is that people who are R users may not be familiar with Java and/or command line so we provide these functions to simplify setting up the OTP.

Currently only tested on Windows 10, should work on other windows version but some functions like otp_stop will not work on other OS.

Simple example:
```
# Create folder "C:/otp" and fill with otp.jar , graphs folder etc
# In R run
otp_build_graph("C:/otp") # Will add the Graph.obj to "C:/otp/graphs/current" folder
otp_setup("C:/otp") # Will run OTP (check with browser)
otp_stop() # Will Kill Java
```

mem48 committed:
Merge pull request #1 from marcusyoung/master

Keeping Up with Marcus

mem48 committed:
Set up functions

Helper functions to remove the need to use the command prompt

mem48 committed:
add stop fucnction and bug fixes

mem48 commented:
P.S. I haven't rebuilt the descriptions etc to avoid merge conflicts

mem48 committed:
minor bug fixes

mem48 committed:
generic routing function

mem48 committed:
little test code

mem48 committed:
demo xml

Robinlovelace commented:
Any thoughts on this @marcusyoung will be good to get it rolling.

marcusyoung commented:
I didn't get a notification from Github about this. I'll take a look.

mem48 committed:
Add elevation support

mem48 committed:
fix typo

mem48 committed:
test elevation

mem48 committed:
Add a sleep of 30 seconds

I takes some time for Java to star up the OTP, but to allow use of R we hand off, and so the function completes instantly. A 30 second wait  prompts the user to wait before visting the OTP.

mem48 committed:
support analyst features

mem48 committed:
support analyist

marcusyoung reviewed (changes requested):
@mem48 

Just an initial thing I spotted. otp-connect has a make_url method which generates the baseurl, so that just needs to be called at line 63.

I will try and look at all your other good stuff as soon as I can.

mem48 committed:
bug fixes

mem48 committed:
elevation improvements

mem48 committed:
working with elevation

mem48 committed:
standardise to <-

mem48 committed:
standardise to <-

mem48 committed:
Fix check_router

If unable to connect the code crashes with an error. Now uses try() so that R does not fail even if there is no connection, and the correct error message is displayed in otp_connect()

mem48 committed:
Make the failure silent

mem48 committed:
otp_setup will lauch OTP

otp_setup will now check the connection using otp_connect(). It will try every 30 seconds for 10 times to connect to the OTP. If it is able to conect it will open OTP in the browser. Needs checking on Linux.

mem48 commented:
@marcusyoung I've switched to make_url() and also made a load more improvements.

1. Now have the otp_route() function which is working but would welcome some testing
2. otp_setup will now check the connection (using otp_connect) and if successful will open OTP in the browser
3. Fixed a bug in check_router.otpconnect() where the code fails rather than returning the expected message.

mem48 committed:
add isochrone support

mem48 committed:
Update Documentation

mem48 committed:
Allow jar in different location

So you don't need many copys of the jar file

mem48 committed:
add demo data

mem48 committed:
Move Data

So it installs correctly

mem48 committed:
fix typo

mem48 committed:
start vignette

mem48 committed:
Support spaces in file paths

mem48 committed:
Support logs

mem48 committed:
Move images

mem48 committed:
Update Vignettes

mem48 committed:
Remove junk

mem48 commented:
I've made some general updates to make it more "package friendly". Adding some vignettes based on your tutorial and sample data. I've also updated the description and documentation so `devtoolls::install_github()` works.

mem48 committed:
support new otp_setup

mem48 committed:
otp_geocode

Convert names to locations

mem48 committed:
error checks

mem48 committed:
fix typo

mem48 committed:
New vignette

mem48 committed:
add data

mem48 committed:
change from character to numeric

mem48 committed:
fix bug

Not sure how this stopped working

mem48 committed:
Fix 3d routing bug

mem48 committed:
Support Batch Routing

Robinlovelace committed:
Update dox 4 linux

Robinlovelace commented:
This PR is getting huge! Looks good, suggest merging soon. Can always tidy-up details after the event.

Biggest problem from CRAN perspective: giant size. 30 MB+. Suggest using the piggyback package or similar to help separate code and data:

```
git clone git@github.com:mem48/opentripplanner-malcolm
Cloning into 'opentripplanner-malcolm'...
remote: Counting objects: 291, done.
remote: Compressing objects: 100% (147/147), done.
remote: Total 291 (delta 114), reused 172 (delta 66), pack-reused 70
Receiving objects: 100% (291/291), 34.04 MiB | 5.79 MiB/s, done.
Resolving deltas: 100% (131/131), done.
```

mem48 committed:
Multi-os support

place holders for supporting different OS types. Currently only windows is working

mem48 commented:
The large size is the sample data, I'd been copying @marcusyoung tutorial.

I slimmed down his original data a bit but still

DEM.tif 5 MB
greater-manchester-osm.pbf 11 MB
tfgm-gtfs.zip 33 MB

Perhaps we should build a new tutorial on a smaller location, e.g. the Isle of White. The problem then will be GTFS data which is waiting on my https://github.com/mem48/UK2GTFS package

marcusyoung commented on R/otp-setup.R:
```diff
@@ -0,0 +1,290 @@
+otp_build_graph <- function(otp = NULL,
+                            dir = NULL,
+                            memory = 2,
+                            router = "current",
```
@mem48

r.e. `otp_build_graph()`

For the help example, as a minimum need to provide:

`otp_build_graph(otp = "C:/otp", dir = "C:/otp")`

If the JAR file is also at the dir location, would it be nice to only have to specify that, and not otp too?

There's an issue with reaching max.print. As the build process tends to produce many  warnings. So I never got to see the message that graph build was ok. Would it be better to pipe the OTP build output to a log file instead?

mem48 replied:
I went back and forth on having a separate otp and dir location. On balance I think a separation of code and data is preferable. If a user really wants them in the same place then having to complete two variables is a small hurdle. The change I have made is to make `otp` a path to the file not the folder, this avoids the problem of a folder containing more than one jar file and makes the difference between the two variables clearer.

I also change the help example to show that the results of the log can be stored

marcusyoung replied:
@mem48 OK, that will make the difference between the two clearer.

marcusyoung commented on R/otp-setup.R:
```diff
@@ -0,0 +1,290 @@
+
+  # Check for errors
+  if(grepl("ERROR",set_up[2])){
+    message("Failed to build graph with message:")
```
I think this message should be \"Failed to start OTP\" or something along those lines.

marcusyoung commented on R/otp-setup.R:
```diff
@@ -0,0 +1,290 @@
+    message("Failed to build graph with message:")
+    message(set_up[2])
+  }
+
```
@mem48

I think this message should be \"Failed to start OTP\" or something along those lines.

mem48 replied:
Not always true, OTP may have started but the build process failed. e.g. missing/corrupt data. So I felt that failed to build graph was more informative.

marcusyoung replied:
@mem48  Sorry I have put the comment in the wrong place. I was referring to `otp_setup()` where the message is the same.

marcusyoung commented on R/otp-setup.R:
```diff
@@ -0,0 +1,290 @@
+                                check = TRUE), silent = T)
+      if("otpconnect" %in% class(otpcon)){
+        message(paste0(Sys.time()," OTP is ready to use Go to localhost:",port," in your browser to view the OTP"))
+        browseURL(paste0(ifelse(otpcon$ssl,"https://","http://"),"localhost:",port))
```
@mem48 Best to use the `make_url` method from `otp_connect()` here.

mem48 replied:
good call

marcusyoung commented on R/otp-setup.R:
```diff
@@ -0,0 +1,290 @@
+  if(!exists("dir")){
+    warning("Path to the files to build graph are not defined")
+    stop()
+  }
```
@mem48

These checks aren't doing as expected. otp and dir will always exist as `otp = null` etc creates an empty object. Want to be checking !is.null() ?

marcusyoung commented on R/otp-setup.R:
```diff
@@ -0,0 +1,290 @@
+      if(java_version < 1.8){
+        warning("OTP requires Java version 1.8 or later")
+        stop()
+      }
```
@mem48

AFAIK JAVA version must be 1.8. It won't work on 9 or 10.

mem48 replied:
According to http://docs.opentripplanner.org/en/latest/Basic-Tutorial/#start-up-otp it is 1.8 or higher. I'm cautious of being too prescriptive as it will mean updating the package each time OTP or Java get an update. It was more intended as a test that the R to Java communication was working.

marcusyoung replied:
That tutorial is incorrect. See issue raised by Robin. Get your point about being too prescriptive but OTP is likely to be very slow in updating for new Java versions. Maybe more important in the Vignettes that it is clear otherwise likely to cause huge frustration.

mem48 replied:
Ok, changed to check that version is 8.X

marcusyoung commented:
Just finished reviewing otp-setup.R. Some great stuff in there. I've made a few comments, see above. Quite a few typos in the help content, not worrying about that at the moment, but will need checking thoroughly at some point.

marcusyoung commented on R/otp-route.R:
```diff
@@ -0,0 +1,287 @@
+  }
+  if(!all(mode %in% c("TRANSIT","WALK","BICYCLE","CAR","BUS","RAIL"))){
+    message("mode is not a valid, can be a character vector of any of these values TRANSIT, WALK, BICYCLE, CAR, BUS, RAIL")
+    stop()
```
Check of mode needs to be a bit more nuanced. Some combinations of these are invalid.

mem48 replied:
uppercase text is easy, what are the rules on which combinations are allowed?

marcusyoung replied:
I'll probably need to dig into the code to get definite answer on that one, though there may be something in the Google Group. It may be best to stick with the options available in the GUI and the modes they generate in the URL (after clicking the view this itinerary link (or whetever it's called). Leave it with me.

marcusyoung commented on R/otp-route.R:
```diff
@@ -0,0 +1,287 @@
+  # Check for errors - if no error object, continue to process content
+  if(is.null(asjson$error$id)){
+    response <- otp_json2sf(asjson, full_elevation)
+    return(response)
```
@mem48

Need to deal with the \'walkTime\' label bug, where it is used for cycle time and drive time too. See line 162 etc. in my original function (in R scripts folder).

mem48 replied:
I would not change this, this is an OTP bug and should be fixed in the OTP, if we start making these changes it will become a nightmare supporting different versions of OTP and moves the package away from being a simple bridge between OTP and R

marcusyoung replied:
Agreed. I will submit bug report on this to OTP (has been around a long time).

marcusyoung commented on R/otp-batch.R:
```diff
@@ -0,0 +1,98 @@
+#' Get multiple routes from the OTP
```
@mem48 I'm not sure this is the way to go with batch processing. This is likely to be a requirement of many of the functions, so it would be better if there was a single internal function which the other functions can call. The user should be able to provide either a single from/to set of coordinates OR a matrix -  in all relevant functions. Otherwise there will be a lot of repetition. Suggest moving this out of the pull request for now or at least not exporting any functions from it.

marcusyoung commented on R/otp-route.R:
```diff
@@ -0,0 +1,287 @@
+#' This function returns a SF data.frame with one row for each leg of the jounrey
+#' (a leg is defined by a change in mode). For transit more than one route option may be returned
+#' and is indicated by the route_option column.
+#'
```
@mem48 Note typo: `@detials`

marcusyoung commented on R/otp-route.R:
```diff
@@ -0,0 +1,287 @@
+#' Get the geometry of a route from the OTP
+#'
```
@mem48 I think we probably need to make clear in the sf column names what the units are for duration, time, distance, elevationGained etc. E.g change labels to distance_km, time_seconds etc. Rather than relying on the user knowing?

marcusyoung commented on R/otp-route.R:
```diff
@@ -0,0 +1,287 @@
+  }else{
+    date <- substr(as.character(date_time),1,10)
+    time <- substr(as.character(date_time),12,16)
+  }
```
@mem248 Having problems get TRANSIT (or BUS) route returned for a specific date. For example, the following on OTP with my Manchester tutorial data, returns trip is not possible error but a valid route is returned using the web GUI, where otpdate is `"2017-10-10 12:15:00 BST"`:

```
otp_plan(otpcon, from = c(53.56172, -2.29202), to = c(53.40892, -2.16637), mode = "BUS", full_elevation = FALSE, date_time = otpdate)
```

This may be due to the date being in the format YYYY-MM-DD when it's extracted from the POSIXct object, while OTP requires MM-DD-YYYY? On that point, I wonder if we really want to require users to provide date and time in POSIXct? It's not that user friendly, and many won't know how to create the object. I had to Google it to remind myself. Think I'd rather provide a basic text format option (or the option of either). What do you think?

mem48 replied:
I know that POSIXct is a little awkward, but it does ensure that we get properly formatted dates. Otherwise, there would be a wide range of formats that would need to be supported or checked.


OTP doesn't seem to support routing in the past, so I would not expect a 2017 date to work

marcusyoung replied:
@mem48 OTP will route based on the schedules provided. If they are in the past then it will route on them.  The schedules in my test instance covered that period and worked in the GUI. But the format must be MM-DD-YYYY in the API request. You can see that in the GUI URL and also see the effect of altering it to YYYY-MM-DD.

marcusyoung replied:
OTP only supports one format though. So as long as the documentation specified:

time as "13:00:00" and date as "MM-DD-YYYY" we would only have to check for that?

marcusyoung replied:
Obviously the date of travel specified in the request must match the transit service validity period (as provided in calendar.txt in the GTFS feed).

marcusyoung replied:
Actually this is one of the great features of OTP (compared to using an online trip planner that will only ever have the current schedules) , you can provide it with old or future data for research. E.g. look at how accessibility has changed due to reduction in bus services over the past 25 years ...

mem48 replied:
I've adjusted the date and time formatting passed to the API, which I think is part of the problem. I can not get historical routing, as far back as the GTFS file goes. Also, are you using the GTFS in the package or your tutorial one? I made some modifications to the one in the package.

I still would prefer the have POSIXct inputs, but perhaps the vignettes can highlight an easy way to make dates e.g. `date_time = lubridate::ymd_hms("2018-04-07 12:15:00")`

marcusyoung replied:
I was using my tutorial GTFS feed for testing. 
That method of making the date is certainly more streamlined!

marcusyoung commented on R/otp-route.R:
```diff
@@ -0,0 +1,287 @@
+
+
+#' Convert Google Encoded Polyline and elevation data into sf object
+#'
```
Is Google relevant here?

mem48 replied:
Yes, Google invented it http://www.danmandle.com/blog/what-is-an-encoded-polyline/

marcusyoung replied:
Oh right. Interesting. I didn't realise OTP used them.

marcusyoung commented on R/otp-geocode.R:
```diff
@@ -0,0 +1,105 @@
+                        stops = TRUE,
+                        clusters = FALSE,
+                        corners = TRUE)
+{
```
type argument not specified. Function fails.

mem48 replied:
fixed, and switched to checkmate checks

marcusyoung commented:
@mem48 Taken a look at otp-geocode. The type argument is missing from the function definition so unable to test.

marcusyoung commented on R/otp-isochrone.R:
```diff
@@ -0,0 +1,111 @@
+                     walkReluctance = 5,
+                     transferPenalty = 0,
+                     minTransferTime = 600,
+                     cutoffSec = c(600,1200,1800,2400, 3000, 3600))
```
@mem48 I don't think CAR should be the default here as it isn't a valid mode for isochrones. It will return polygons but they won't be correct. Need to think about whether we even allow non-transit modes as options here.

marcusyoung commented:
@mem48 Have taken a look at otp-isochrone. This has similar problem to otp-route for TRANSIT modes when specifiying date and time, I think because the date format is not right for OTP. Currently returns empty polygons in this situation, so also need to look at how we pick up errors.

Function currently returns only an SF object. I think going forward the return format should be optional, additionally including GEOJSON (and possibly SHAPEFILE) Both are the native options available from OTP. Users will not always want to continue processing in R, e.g. writing direct to PostgreSQL.

marcusyoung commented on vignettes/dummies_guide.Rmd:
```diff
@@ -0,0 +1,137 @@
+
+* Download free Java 8 from https://www.java.com/en/download/ and install Java. You will need the correct version for your operating system.
+* Download the .shaded.jar OTP  file from https://repo1.maven.org/maven2/org/opentripplanner/otp/ at time of writing the latest version of OTP is version 1.3 so you would want to get https://repo1.maven.org/maven2/org/opentripplanner/otp/1.3.0/otp-1.3.0-shaded.jar 
+
```
@mem48 Probably need a word of warning here. The release version of OTP is not well-suited for all counries (it is tuned for USA). The OTP.JAR in the tutorial is a modified version where I change edge traversal permissions and adjusted speed limits to better match the UK situation. I think that this may now be possible to configure via the new Way property sets that can be configured and are then read at graph build. I will investigate further. This info can then be incorporated into the advanced guide (probably?).

marcusyoung commented on vignettes/advanced_features.Rmd:
````diff
@@ -0,0 +1,108 @@
+
+```{r, echo = FALSE, fig.align='center', fig.cap="\\label{fig:otpgui}Isochrones from Manchester Airport"}
+knitr::include_graphics("images/isochrone.png")
+```
````
@mem48 There's something not right about these isochrones in the map. Full of holes. Compare with the one in the tutorial. It may due to WALK being included in the modes. Isochrones only work correctly for TRANSIT modes.

marcusyoung replied:
Actually, given the other issue regarding date format mention above, I suspect this is actually the WALK isochrone for Manchester Airport as no Transit would have been returned. But WALK won't returrn a proper valid isochrone, thus all the holes.

marcusyoung commented:
@mem48 Have had a look at the new Vignettes. Great work. A few comments above.

marcusyoung commented:
@mem48 Looks like we've got some internal functions appearing in the man folder.

mem48 committed:
Direct to JAR file

mem48 commented on R/otp-setup.R:
```diff
@@ -0,0 +1,290 @@
+      warning(paste0("Unable to find Graph.obj in ",dir,"/graphs/",router))
+      stop()
+    }
+  }
```
done

mem48 committed:
make fixes

mem48 commented on R/otp-route.R:
```diff
@@ -0,0 +1,287 @@
+#' Get the geometry of a route from the OTP
```
The function basically gets everything that OTP returns, that seems a good default. We could add a variable to only return the metadata, which would provide a small performance boost

marcusyoung replied:
That would be useful I think.

mem48 replied:
ok I will add

mem48 committed:
more check fixes

mem48 committed:
fix checks

mem48 committed:
fix checks

mem48 committed:
fix checks

mem48 committed:
fix checks

mem48 committed:
Support NA produced from JAVA 9

marcusyoung commented:
Replaced by pr17
